File: src/exemplos/05_EstruturasDeDados/06_Arquivo/01-binario.c
/**      @file: 01-binario.c
 *     @author: Guilherme N. Ramos (gnramos@unb.br)
 * @disciplina: Algoritmos e Programação de Computadores
 *
 * Exemplo de leitura e escrita em arquivo binário. */

#include <stdio.h>
#include <stdlib.h>

#define LETRAS 26 /* Valor constante global. */
#define NUMEROS 12 /* Valor constante global. */

/* Cria um arquivo com o nome dado como argumento. Retorna EXIT_SUCCESS caso
 * seja possível, EXIT_FAILURE caso contrário. */
int cria(char* arquivo) {
    const char string[] = "abcdefghijklmnopqrstuvwxyz";
    int i;
    FILE *fp = fopen(arquivo, "wb+");

    if(!fp) {
        printf("Não foi possível abrir \"%s\".\n", arquivo);
        return EXIT_FAILURE;
    }

    /* Escreve string em bloco. */
    fwrite(string, sizeof(string), 1, fp);

    /* Escreve números, um de cada vez. */
    for(i = 1; i <= NUMEROS; ++i)
        fwrite(&i, sizeof(i), 1, fp);

    fclose(fp);

    return EXIT_SUCCESS;
}

/* Mostra na saída padrão o conteúdo do arquivo dado como argumento. */
void mostra(char* arquivo) {
    char c;
    int i, numeros[NUMEROS];

    FILE *fp = fopen(arquivo, "rb");

    if(fp != NULL) {
        /* Lê um caractere de cada vez, +1 para o caractere
         * de terminação do string.  */
        printf("string: ");
        for(i = 0; i < LETRAS + 1; ++i) {
            fread(&c, sizeof(c), 1, fp);
            putchar(c);
        }

        /* Lê em bloco. */
        printf("\nnumeros:");
        fread(numeros, sizeof(numeros), 1, fp);

        for(i = 0; i < NUMEROS; ++i)
            printf(" %d", numeros[i]);
        printf("\n");

        fclose(fp);
    }
}

int main() {
    char arquivo[] = "apc.c.bin";

    if(cria(arquivo) == EXIT_SUCCESS)
        mostra(arquivo);


    return EXIT_SUCCESS;
}

File: src/exemplos/05_EstruturasDeDados/06_Arquivo/01-binario.py
#  -*- coding: utf-8 -*-
#    @package: 01-binario.py
#     @author: Guilherme N. Ramos (gnramos@unb.br)
# @disciplina: Algoritmos e Programação de Computadores
#
# Exemplo de leitura e escrita em arquivo binário.


import struct


LETRAS = 26  # Valor constante global.
NUMEROS = 12  # Valor constante global.


def cria(arquivo):
    '''Cria um arquivo com o nome dado como argumento. Retorna True caso seja
    possível, False caso contrário.
    '''
    string = 'abcdefghijklmnopqrstuvwxyz'

    try:
        with open(arquivo, 'wb') as f:
            # Escreve string em bloco.
            f.write(struct.pack('26s', string.encode('ASCII')))

            # Escreve números, um de cada vez.
            for i in range(1, NUMEROS + 1):
                f.write(struct.pack('i', i))

            return True
    except:
        print('Não foi possível abrir \'%s\'.\n' % arquivo)
        return False


def mostra(arquivo):
    '''Mostra na saída padrão o conteúdo do arquivo dado como argumento.'''
    try:
        with open(arquivo, 'rb') as f:
            # Lê um caractere de cada vez, +1 para o caractere de terminação do
            # string.
            string = ''
            for i in range(LETRAS):
                string += f.read(1).decode('ASCII')
            print('string: ' + string)

            # Sabendo que o tamanho padrão de um inteiro (escrito no arquivo
            # com struct.pack('i', i), é possível ler em bloco com o comando:
            # f.read(4*NUMEROS)
            #
            # Mas isso resulta em um string binário que precisa ser processado.
            # Uma alternativa mais simples é ler de um em um:
            numeros = [struct.unpack('i', f.read(4))[0] for _ in range(NUMEROS)]
            print('numeros:', numeros)
    except:
        print('Não foi possível abrir \'%s\'.\n' % arquivo)


if __name__ == '__main__':
    arquivo = 'apc.python.bin'

    if cria(arquivo):
        mostra(arquivo)
